From 7ef2d3c8dfc09a41a5b324966e58d1cdaee9614a Mon Sep 17 00:00:00 2001
From: "Cao,Bing Bu" <bingbu.cao@intel.com>
Date: Wed, 13 May 2015 16:32:44 +0800
Subject: [PATCH 396/462] atomisp: qbuf return error if get info failed

In rare cases, camera3hal may queue buffer before stream
created or after stream destroyed, the driver should
check the return value of the ia_css_pipe_get_info(),
further buffer map and queue ops should NOT be allowed.

This patch fix this issue.

Change-Id: I7d5281980b6b4462e634a11f7ac5372edca988b1
Tracked-On: https://jira01.devtools.intel.com/browse/IMINAN-35864
Signed-off-by: Cao,Bing Bu <bingbu.cao@intel.com>
Reviewed-on: https://icggerrit.ir.intel.com/55780
Build: IMIN (SC)
Approver: Jing Tao <jing.tao@intel.com>
Reviewed-by: Yunliang Ding <yunliang.ding@intel.com>
Integrator: WeiX Meng <weix.meng@intel.com>
Build: WeiX Meng <weix.meng@intel.com>
Reviewed-by: WeiX Meng <weix.meng@intel.com>
Maintainer: WeiX Meng <weix.meng@intel.com>
---
 .../media/pci/atomisp2/atomisp_driver/atomisp_compat_css20.c  | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_compat_css20.c b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_compat_css20.c
index e431cfb..1cb8ef2 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_compat_css20.c
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_compat_css20.c
@@ -3009,6 +3009,8 @@ int atomisp_get_css_frame_info(struct atomisp_sub_device *asd,
 	struct ia_css_pipe_info info;
 	int pipe_index = atomisp_get_pipe_index(asd, source_pad);
 	int stream_index;
+	struct atomisp_device *isp = asd->isp;
+
 	if (ATOMISP_SOC_CAMERA(asd))
 		stream_index = atomisp_source_pad_to_stream_id(asd, source_pad);
 	else {
@@ -3016,8 +3018,13 @@ int atomisp_get_css_frame_info(struct atomisp_sub_device *asd,
 			   ATOMISP_INPUT_STREAM_VIDEO :
 			   atomisp_source_pad_to_stream_id(asd, source_pad);
 	}
-	ia_css_pipe_get_info(asd->stream_env[stream_index]
-		.pipes[pipe_index], &info);
+
+	if (IA_CSS_SUCCESS != ia_css_pipe_get_info(asd->stream_env[stream_index]
+				 .pipes[pipe_index], &info)) {
+		dev_err(isp->dev, "ia_css_pipe_get_info FAILED");
+		return -EINVAL;
+	}
+
 	switch (source_pad) {
 	case ATOMISP_SUBDEV_PAD_SOURCE_CAPTURE:
 		*frame_info = info.output_info[0];
-- 
1.9.1

